Adding Multispeed Coil Speed Level Control Actuators

xuanluo113 opened the pull request:
Pull request overview
---------------------
This PR makes the speed level of the cooling and heating multispeed coils, namely `Coil:Cooling:DX:MultiSpeed` and `Coil:Heating:DX:Multispeed`, actuable. Both the multispeed cooling and heating coil can only be referenced by an `AirLoopHVAC:UnitarySystem` or `AirLoopHVAC:UnitaryHeatPump:AirToAir:Multispeed` object.

Currently, both `AirLoopHVAC:UnitarySystem` or `AirLoopHVAC:UnitaryHeatPump` have a reportable variable “Unitary System DX Coil Speed Level” indicating the index of the speed level, and the variable is originally controlled either by load or temperature. We added EMS actuators to both speed level indices of the unitary system and unitary heat pump and overwrite the values of both variables at the end of the control logic. 

We added two test files to test the new actuator with both the traditional EMS program and the Python Plugin System.

The audit diffs are expected due to additions to available actuators.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

xuanluo113 committed:
NFP

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
unitary heat pump

nrel-bot-2b commented:
@xuanluo113 @lgentile it has been 28 days since this pull request was last updated.

xuanluo113 committed:
temp

xuanluo113 committed:
clean up multispeed heat pump

xuanluo113 committed:
fix a bug

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
unitary system

xuanluo113 committed:
unitary and unit test

xuanluo113 committed:
unitary fix and add a unit test

xuanluo113 committed:
add python plugin test files

xuanluo113 committed:
documentation

xuanluo113 commented:
@lgu1234 @Myoldmopar @mjwitte @EnergyArchmage This branch is ready for review. CI is clean and tests and documentation are updated. @lgu1234 Would you mind reviewing this branch as it's related to the coil speed control? Mike, Edwin and Brent, if you have time would you also take a look as you have the context of this branch in our previous discussions? Please also feel free to suggest reviewers you see as the best fit. Thanks!

lgu1234 commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-001/coils.tex:
```diff
@@ -2604,6 +2604,10 @@ \subsubsection{Seasonal Energy Efficiency Ratio (SEER) for Multi-Speed DX Coil}\
 
+\subsubsection{Energy Management System Coil Speed Level Override}\label{energy-management-system-coil-speed-level-override-1}
+
+The cooling coil speed level could be actuated by EMS to at run-time. When EMS is used to overwrite the “Unitary System DX Coil Speed Level” values, the speed levels are applied to the EMS values rather than the default controlled speed levels.
```
The speed level is calculated at every iteration based on load and system capacity. It is not default. Here is my suggestion:

the speed levels are applied to the EMS values rather than the load based speed levels.

xuanluo113 replied:
Other than the load-based control logic, there is also setpoint-based control as alternative control logic in the native EnergyPlus program. Would it be clearer to say "the speed levels are applied to the EMS values rather than speed levels determined by the native EnergyPlus control logic"? Thanks!

lgu1234 replied:
I think that EMS value will be native after this release. How about Non-EMS control logic, instead of native?

xuanluo113 replied:
👍

lgu1234 commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-001/coils.tex:
```diff
@@ -2770,6 +2774,10 @@ \subsubsection{Standard Ratings}\label{standard-ratings}
 
+\subsubsection{Energy Management System Coil Speed Level Override}\label{energy-management-system-coil-speed-level-override-2}
+
+The heating coil speed level could be actuated by EMS to at run-time. When EMS is used to overwrite the “Unitary System DX Coil Speed Level” values, the speed levels are applied to the EMS values rather than the default controlled speed levels.
```
See above comment

lgu1234 commented on src/EnergyPlus/HVACMultiSpeedHeatPump.cc:
```diff
@@ -172,6 +172,15 @@ namespace HVACMultiSpeedHeatPump {
+                                 "[]",
+                                 state.dataHVACMultiSpdHP->MSHeatPump(coilNUM).EMSOverrideCoilSpeedNumOn,
+                                 state.dataHVACMultiSpdHP->MSHeatPump(coilNUM).EMSOverrideCoilSpeedNumValue);
+            }
```
I looked at DXCoils module. It uses
    if (state.dataGlobal->AnyEnergyManagementSystemInModel) {  
before SetupEMSActuator.

If there are no EMS inputs in an idf file, is it necessary to call this section?

lgu1234 commented on tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc:
```diff
@@ -1401,6 +1401,23 @@ TEST_F(EnergyPlusFixture, HVACMultiSpeedHeatPump_ReportVariableInitTest)
+    EXPECT_NEAR(state->dataLoopNodes->Node(22).HumRat, 0.008, 0.0001);
+    EXPECT_NEAR(state->dataLoopNodes->Node(22).Enthalpy, 47077.4613, 0.0001);
+    EXPECT_NEAR(state->dataHVACMultiSpdHP->MSHeatPump(2).CompPartLoadRatio, 0.1530992, 0.0001);
+
```
I tested the unit test. It worked as expected. When I removed EMS overrides (lines 1405, 1406 and 1413), the unit test still passes. That means the override values are the same as calculated one. In other words, the EMS override does not override calculated values.

Then I set up value = 1 on line 1406 and value = 2 at line 1413. Then speedNum = 1 and CycRatio = 1 in the first case, as expected. However, the second case provide a fatal message as "DX unit compressor speed calculation failed: speed limits exceeded...." In other words, when the override value is greater than the calculated value, no solution will be provided.

Here is my suggestion.

1. Allow the program to calculate SpeedNum also.
2. Compare EMS override value with SpeedNum
3. If EMS > SpeedNum, use calculated SppedNum, and provide warning: Override Value will not be used, Simulation continues...
4. If EMS < = SpeedNum, use EMS value

xuanluo113 replied:
1. The test case does successfully override the calculations, as when `EMSOverrideCoilSpeedNumOn = true;` in line 1406, the program used the [EMS control branch](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/src/EnergyPlus/HVACMultiSpeedHeatPump.cc#L310) instead of the original branch.

2. When removing lines 1405 and 1406, the unit test still passes, because without EMS override, the program redirection to normal control [at here](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/src/EnergyPlus/HVACMultiSpeedHeatPump.cc#L358), and calculated speed num under `QZnReq = 50000.00;` is also 2, same as the EMS override case. And under the same speed level, the `CycRatio` and `SpeedRatio` are supposed be the same.

xuanluo113 replied:
3. [Line 1413 to 1419 ](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc#L1419) tests that the EMS override logic reproduces the same system partial load output with the normal control. With SpeedNum set back to 1 by EMS override value, the partial load ratio is the same  with the normal control calculated [partial speed load ratio](https://github.com/NREL/EnergyPlus/blob/112aca098677382a21689d5c2f32942559b81e31/tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc#L1401).

xuanluo113 replied:
4. I'm not sure about the logic that if the EMS override SpeedNum is not solvable, we switch back to the normal control. On the one hand, the user-defined EMS program should take full control and should be responsible for making sense of the control logic. If the EMS override value is not solvable, it should report an error. On the other hand, I also have the feeling that it's hard for an external program to achieved a solvable logic with fewer handlers than the defined control program in E+, so it might be fairly easy for the control to reach an unsolvable state. @Myoldmopar @mjwitte Edwin and Mike, were there similar issues in Plant Loop override control as well? How was the problem solved there?

lgu1234 replied:
Here is my thought about Item 4.

If we know that the EMS override is not solvable, why do we need override?

xuanluo113 replied:
Because I feel like that's what the user requests, to change the logic. If the request is not legitimate, then they should be notified to modify their desired logic. I think users would only use this feature if they want to override the Non-EMS logic, and overriding their overrides would be a bit confusing.

EnergyArchmage replied:
I would expect EMS override to be done using short timesteps, with logic based on temperature differences, eg. if zone temp is off by 1.0C use speed 1, if off by 2.0C use speed 2.  The part load with in a speed level is not needed. Just always use part load of 1.0 with EMs override of speed number and run. nothing to solve.  Control will be realistic but different than usual energyplus load-based, long timestep approach.

xuanluo113 commented:
@JasonGlazer @mjwitte I have a question about the reporting of the coil speed level in the ESO file with HVAC Timesteps. I'm not sure if it's my misunderstanding of the system time step or there's an in the program, but the reportable `Unitary System DX Coil Speed Level` seems to be one HVAC time step behind the run time printed calculation values.

For example, Figure 1 below is the run time printed system variables and heat pump speed levels during the coil speed calculation, listing the first 5 HVAC time steps (SysTimeStep = 1 min here so the first 5 minutes in design day 1/21). There seem to be four iterations of coil calculations per system time step. The variables are printed at the end of `SolveAirLoopControllers` , after the coil speed and load ratios are resolved.

<img width="602" alt="Screen Shot 2021-08-17 at 10 49 31 AM" src="https://user-images.githubusercontent.com/23508095/129788041-a7a37754-6d58-4f5c-92bc-fc19f8de5149.png">

At 00:00 am, the speed level was solved as 2. At 00:01 am, the speed level is set to 1, while at 00:02 am, the speed level is set to 2. 

However, in figure 2 (screen shot of the CSV output) the reporting of the coil speed seems to be different from run time printouts. There's no reporting for sys time step 00:00 am, and from 00:01 am the reported values seem to be lagging one time step behind.
<img width="439" alt="Screen Shot 2021-08-17 at 10 33 24 AM" src="https://user-images.githubusercontent.com/23508095/129788658-7333b165-dcf1-4be3-94cb-0349bc2d0d06.png">

I don't think how reportable variables at system timesteps are aggregate has anything to do with the scope of this PR. But I wonder if there's a potential issue in this?

mjwitte commented:
@xuanluo0113 The variable for Unitary System DX Coil Speed Level is `m_SpeedNum`. Based on a quick search through the code for `m_EMSOverrideCoilSpeedNumValue` it appears that this sets either `m_HeatingSpeedNum `, `m_CoolingSpeedNum`, or `m_SpeedNum`. Then in `reportUnitarySystem` there's this:
`this->m_SpeedNum = max(this->m_CoolingSpeedNum, this->m_HeatingSpeedNum);`

Perhaps the problem is with the 2 places that set `this->m_SpeedNum = this->m_EMSOverrideCoilSpeedNumValue`? I think they should be setting either `m_HeatingSpeedNum ` or `m_CoolingSpeedNum`. Just guessing.

xuanluo113 commented:
@mjwitte Thanks. If the comment is about the reported issue, the mismatch speed level is of the multi-speed heat pump, `state.dataHVACMultiSpdHP->MSHeatPumpReport(coilNUM).SpeedNum`. Sorry for not making it clear.

And I just realized that the behavior is normal because until the end of `SimHVAC`, the system zone air temperatures are not updated to the current time step yet, and all coil controls or calculations are based on the last time step's zone condition. I got confused about the causes and the results in the reported variables.

<img width="418" alt="Screen Shot 2021-08-18 at 11 29 54 AM" src="https://user-images.githubusercontent.com/23508095/129953314-c2ccf390-0a0a-4814-82bb-975103c96a92.png">

xuanluo113 committed:
merge develop, update documenation and error messages

xuanluo113 committed:
test file add space

xuanluo113 commented:
The comments so far are either addressed or replied to. This branch is ready for another review.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
[skip ci] temp

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

Myoldmopar commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -1538,7 +1537,18 @@ namespace UnitarySystems {
+                                 state.dataUnitarySystems->unitarySys[coilNUM].m_EMSOverrideCoilSpeedNumOn,
+                                 state.dataUnitarySystems->unitarySys[coilNUM].m_EMSOverrideCoilSpeedNumValue);
+            }
+        }
```
@rraustad @mjwitte how do you feel about this in relation to the new coil?  I'm not necessarily saying this branch needs to change, but we should definitely be aware that this is yet another thing being added to one specific coil model.

rraustad replied:
Well, 1) this belongs in setupAllOutputVars, not here. And 2), it should apply to all coil types that use `this->m_CoolingSpeedNum` so it just needs to be tested with the new coil model to make sure it works. MultiSpeed is different from VS so it likely won't work the first time. This conditional (if Multi*) does prevent that from causing problems with the release so that's good. A follow up issue could expand this to other coil types.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

Myoldmopar commented on testfiles/CMakeLists.txt:
```diff
@@ -248,6 +248,7 @@ add_simulation_test(IDF_FILE EMSCustomOutputVariable.idf EPW_FILE USA_IL_Chicago
 add_simulation_test(IDF_FILE EMSCustomSchedule.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE EMSDemandManager_LargeOffice.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE EMSDiscreteAirSystemSizes.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+add_simulation_test(IDF_FILE EMSMultiSpeedHeatPumpOverrideControl.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
```
:+1:

Myoldmopar commented on testfiles/CMakeLists.txt:
```diff
@@ -436,6 +437,7 @@ if (LINK_WITH_PYTHON)
     add_simulation_test(IDF_FILE PythonPluginCustomSchedule.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
     add_simulation_test(IDF_FILE PythonPluginDemandManager_LargeOffice.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
     add_simulation_test(IDF_FILE PythonPluginDiscreteAirSystemSizes.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+    add_simulation_test(IDF_FILE PythonPluginMultiSpeedHeatPumpOverrideControl.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
```
:+1:

Myoldmopar reviewed (commented):
I feel like this is in pretty good shape.  It doesn't make me feel _great_ that it looks like it is just adding capability to a single coil type, and not even the new Coil:Cooling:DX.  (Please correct me if I'm wrong.)  Other than that I'm OK with the changes, and happy to see matching EMS and Python example files.  Others had different review comments regarding the operation, so I will defer to them if there are other functional changes to make.

xuanluo113 committed:
remove solving partial load ratio

xuanluo113 commented:
@EnergyArchmage 

Brent, I just modified the code based on your latest comment (and what folks suggested during last Wednesday's iteration call), by removing the `SolveRoot` functions to solve partial load ratios with user input speed levels. Currently, with EMS overwrite of the speed level, when `SpeedNum` is one, the multi-speed HP is calculated at ` PartLoadFrac = 1.0` and `SpeedRatio = 0.0`, and when `SpeedNum > 1`, the multi-speed HP is calculated at ` PartLoadFrac = 1.0` and `SpeedRatio = 1.0`. [Most changes in this commit](https://github.com/NREL/EnergyPlus/pull/8865/commits/19556fd80d6c79abd4e5b7ab667ffe36316fbd71).

I wanted to double-check if my understanding of the logic is right here. And I was also wondering if the supplementary heater logic should follow after the HP calculation, as in the non-overwrite control.

I also modified the `testfiles/EMSMultiSpeedHeatPumpOverrideControl.idf` EMS program to include a more realistic overriding control logic, based on heating and cooling set points. Screenshots of sample control results in design days are attached.

<img width="884" alt="Screen Shot 2021-09-07 at 12 14 44 PM" src="https://user-images.githubusercontent.com/23508095/132400013-60ee218a-2c2c-4c84-bdd0-93c66608afdb.png">

<img width="656" alt="Screen Shot 2021-09-07 at 12 14 52 PM" src="https://user-images.githubusercontent.com/23508095/132400043-80963f3d-f3a6-4727-bf61-11a966942073.png">

xuanluo113 commented:
@Myoldmopar Based on the NFP, this PR "proposes to make the speed level of the cooling and heating multispeed coils, namely `Coil:Cooling:DX:MultiSpeed` and `Coil:Heating:DX:Multispeed`, actuable. Both the multispeed cooling and heating coil can only be referenced by an `AirLoopHVAC:UnitarySystem` or `AirLoopHVAC:UnitaryHeatPump:AirToAir:Multispeed` object." Are these not parts of the new coil? 

Both `AirLoopHVAC:UnitarySystem `or AirLoopHVAC:UnitaryHeatPump have a reportable variable “Unitary System DX Coil Speed Level” and here we added EMS actuators to both speed level indices of the unitary system and unitary heat pump.

Myoldmopar commented:
@xuanluo0113 that's what I figured.  The `Coil:Cooling:DX:MultiSpeed` is one of the coils that is tabbed for removal once the `Coil:Cooling:DX` object gets full coverage in parent objects.  The original scope just should have read: 

> `Coil:Cooling:DX:MultiSpeed` and `Coil:Heating:DX:Multispeed` and _multi-speed_ `Coil:Cooling:DX` objects

Anyway, if this gets fully wrapped up for just the MultiSpeed coils then so be it, it can drop in, but it will just be another thing that is different between our coil objects.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
address comments

Myoldmopar commented:
I think this is actually quite clean.  The diff appears to be due to the branch being behind develop.  And the one test failure seems to be an anomaly.  If this was otherwise fully ready I'd say let's get another clean CI pass.  But if there is more to do then I won't worry about that.

xuanluo113 commented:
@Myoldmopar I'm actually still working on the idea to calculate the Cycling / Speed ratio based on user input continuous speed level. This needs another round.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
use continuous speed value

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 committed:
test file

Myoldmopar commented:
CI results:
 - Linux Release: Dataset fixture failure is unrelated, ERR diff is unrelated (both fixed in #9055).  Just audit diffs.
 - Mac: Same, just audit diffs
 - Linux unit tests, just the unrelated unit test failure -- all clean
 - Linux integration: all clean
 - Windows: unrelated dataset fixutre, and the GPU integration failure on the new Windows CI bot -- all clean.

Looks good from a CI standpoint, just some audit diffs.  @xuanluo0113 can you go through the last code changes and provide a little walkthrough of what you did so that we understand it?  This can then drop in if no issues are identified.

xuanluo113 committed:
address comments for unitary hp

xuanluo113 committed:
clean up unitary system

xuanluo113 committed:
clean up

xuanluo113 commented:
@Myoldmopar Thanks for the CI checks. I just pushed a few clean-up commits on docs, error messages, and tests.

The code change since the last technicalities call includes:
- Changed the discrete speed level number overwrite to continuous speed value overwrite (to avoid preassuming cycling ratio or solving the ratio). As updated in the coil docs, "The EMS override coil speed value is a continuous number below the maximum coil speed level allowed. With any EMS override coil speed value, the speed level is calculated as the closest integer greater than or equal to the EMS speed value (the ceiling), and the cycling/speed ratio is set as the floating-point part of the EMS speed value. For example, if the EMS-overwritten coil speed value is 1.2, the speed level number is set as 2 with a speed ratio of 0.2. If the EMS-overwritten value is 0.15, the speed level is set as 1 with a cycling ratio of 0.15).
- The unit test `ZoneUnitarySysTest.UnitarySystemModel_MultiSpeedDXCoolCoil_Only` was modified to test the continuous speed number logic. The test first runs a coil calculation without EMS and records the speed number (e.g. 1) and cycling ratio (e.g. 0.708128) at the certain load. The test then uses `0.708128` as the `EMSOverrideCoilSpeedNumValue` and test if the coil calculation generates the same results in terms of speed level, cycling ratio, and system node condition.
- Similar code and test changes were made in `HVACMultiSpeedHeatPump`.
- Added the error that the ceiling of EMS override value should not be larger than the maximum coil speed allowed.
- Separated the `controlUnitarySystemOutputEMS` logic out of `controlUnitarySystemOutput`.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 commented:
@Myoldmopar The CI looks clean in this branch. Could you take another look at the updated logic explained in this thread and the documentation?

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

mjwitte commented on src/EnergyPlus/HVACMultiSpeedHeatPump.cc:
```diff
@@ -2956,6 +3015,179 @@ namespace HVACMultiSpeedHeatPump {
+        OnOffAirFlowRatio = 0.0;
+        SupHeaterLoad = 0.0;
+
+        // Get full load result
+        PartLoadFrac = 1.0;
+        SpeedRatio = 1.0;
```
I don't see why the full output section is needed here.

xuanluo113 replied:
As discussed below.

mjwitte commented on src/EnergyPlus/HVACMultiSpeedHeatPump.cc:
```diff
@@ -2956,6 +3015,179 @@ namespace HVACMultiSpeedHeatPump {
+
+        if ((QZnReq > SmallLoad && QZnReq > FullOutput)) {
+            Real64 TempOutput;
+            PartLoadFrac = 1.0;
+            SpeedRatio = 1.0;
```
So, `ControlMSHPSupHeater` can change PartLoadFrac and SpeedRatio here? Won't that step on the EMS override value?

xuanluo113 replied:
As discussed below.

mjwitte commented on src/EnergyPlus/HVACMultiSpeedHeatPump.cc:
```diff
@@ -2956,6 +3015,179 @@ namespace HVACMultiSpeedHeatPump {
+                              int const SpeedNum,            // Speed number
+                              Real64 &SpeedRatio,            // unit speed ratio for DX coils
+                              Real64 &PartLoadFrac,          // unit part load fraction
+                              Real64 &OnOffAirFlowRatio,     // ratio of compressor ON airflow to AVERAGE airflow over timestep
```
Shouldn't these be constants here? If any of these change, then the coil simulation in `ControlMSHPOutputEMS` won't be valid. What am I missing here?

xuanluo113 replied:
As discussed below.

mjwitte commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -8138,6 +8137,126 @@ namespace UnitarySystems {
+
+        if (state.dataUnitarySystems->HeatingLoad) {
+            if (SpeedNumEMS > this->m_NumOfSpeedCooling) {
+                ShowFatalError(state,
```
Users may find these fatal errors to be annoying. It's fine for now, but it may be better to change these to a one-time warning and just reset to the max speed num. Same up in MSHP.

xuanluo113 replied:
I don't know about this. If the user creates a 4-speed coil but intendedly overrides the speed level to 5, shouldn't it be an error to fix (explicitly)?

rraustad replied:
I think you should report a recurring warning and operate at the maximum speed level/ratio.

xuanluo113 replied:
Okay. Working on it.

mjwitte commented on testfiles/EMSMultiSpeedHeatPumpOverrideControl.idf:
```diff
@@ -0,0 +1,2040 @@
+  EnergyManagementSystem:Program,
+    CoilSpeedControl,                                                                            !- Name
+    IF (ZoneAirTemperature < HeatingSetpoint),                                                   !- Program Line 1
+    SET CoilSpeedOverride = 1.9,                                                                 !- Program Line 2
+    SET CoilSpeedLevelOverrideReport = 1,                                                        !- Program Line 3
```
I would expect CoilSpeedLevelOverrideReport =2 here. Same a few lines down.

mjwitte replied:
Oh, I see, I misunderstood the meaning of the variable. Never mind.

mjwitte commented on testfiles/PythonPluginMultiSpeedHeatPumpOverrideControl.py:
```diff
@@ -0,0 +1,121 @@
+            heating_setpoint = self.api.exchange.get_variable_value(state, self.heating_setpoint_handle)
+            cooling_setpoint = self.api.exchange.get_variable_value(state, self.cooling_setpoint_handle)
+            if zone_air_temp < heating_setpoint - 2:
+                self.api.exchange.set_actuator_value(state, self.coil_speed_level_handle, 2.0)
+                self.api.exchange.set_global_value(state, self.coil_speed_override_report_handle, 1.0)
```
report = 2 here?

mjwitte commented on testfiles/PythonPluginMultiSpeedHeatPumpOverrideControl.py:
```diff
@@ -0,0 +1,121 @@
+            elif zone_air_temp < heating_setpoint:
+                self.api.exchange.set_actuator_value(state, self.coil_speed_level_handle, 1.0)
+                self.api.exchange.set_global_value(state, self.coil_speed_override_report_handle, 1.0)
+            elif zone_air_temp > cooling_setpoint:
+                self.api.exchange.set_actuator_value(state, self.coil_speed_level_handle, 1.0)
```
Is the intent is for the python version to be the same as the EMS version? The python controls are using different speed levels and never use cooling speed 2.

xuanluo113 replied:
Modified them to be consistent.

mjwitte reviewed (commented):
Tested the example files, no problems there.
Ran an annual simulation, then changed the program to always set the speeds to 1.0: less energy use and more setpoint not met hours. It works!

Unit tests run serially pass.

Most important item: The new actuator needs to be added to the EMS Application Guide.
This can drop in as-is today, as long as the doc change goes in before release.
Same for idf edits.

Other comments can be addressed later, if needed.

mjwitte commented:
Actually, looking more closely at the results for EMSMultiSpeedHeatPumpOverrideControl (run as-is with design days):
1. HEAT PUMP DX HEATING COIL 1:Heating Coil Heating Energy is always zero.
2. The supp heat coil is doing all the work.
3. The cooling results make sense.
4. I changed the EMS logic to always set the speed = 1.0, and the coils never run and the supp heat load is zero??
[EMSMultiSpeedHeatPumpOverrideControl-ChangeSpeeds.idf.txt](https://github.com/NREL/EnergyPlus/files/7186115/EMSMultiSpeedHeatPumpOverrideControl-ChangeSpeeds.idf.txt)

xuanluo113 committed:
integer value ratio fix

xuanluo113 commented:
@mjwitte Thanks for the review. I'll update the EMS App Guide and Python API example file asap, but to answer your latest comment:
- I made a mistake in calculating the cycling ratio with Integer speed value input, when input `SpeedVal` is a integer, `ceil(SpeedVal) == floor(SpeedVal) == SpeedVal` and the cycling/speed ratio should be set at 1.0 here. (e.g. Input Speed = 1.0, `SpeedLevel = ceil(SpeedVal) = 1.0`, and `Cyc/SpeedRatio = 1.0`.
- But even if I modified this, the coils are still all off in heating mode because `OutsideDryBulbTemp < MSHeatPump(MSHeatPumpNum).MinOATCompressorHeating`, and [after this line of check](https://github.com/NREL/EnergyPlus/blob/f148d65cd4c970438777a8914021b2410925eef8/src/EnergyPlus/HVACMultiSpeedHeatPump.cc#L3897), the coils are simulated at speed/prl = 0.0 [here](https://github.com/NREL/EnergyPlus/blob/f148d65cd4c970438777a8914021b2410925eef8/src/EnergyPlus/HVACMultiSpeedHeatPump.cc#L3920) (as all off).

mjwitte commented:
Ok, so I should see heating coil operating during the annual simulation. By the way, it would be good to remove the reportSch from the output variables in the example idfs (it's off for the entire annual simulation - or edit the reportschedule.)
Remember to mention in the EMS actuator description that the temperature limit and overall cooling/heating load = zero will shut off the coil regardless of the EMS speed setting.

xuanluo113 commented:
@mjwitte And for the supplementary heater, I have to calculate the `FullOutPut` load because the `SupHeaterLoad` is based on the residual of `QZnReq - FullOutput`, and when that happens, it seems the `PartLoadFrac = 1.0; SpeedRatio = 1.0` and reevaluated `OnOffAirFlowRatio` should be local for the supplementary heat calc (as hard reset to simulate full load)?

mjwitte commented:
> And for the supplementary heater, I have to calculate the FullOutPut load because the SupHeaterLoad is based on the residual of QZnReq - FullOutput, and when that happens, it seems the PartLoadFrac = 1.0; SpeedRatio = 1.0 and reevaluated OnOffAirFlowRatio should be local for the supplementary heat calc (as hard reset to simulate full load)?

Well, under normal operation that would be true, but it seems to me for this purpose the sup heater load should be `QZnReq - TempOutput` (at least for continuous fan). I don't know how this all goes together for cycling fan operation. @rraustad ???

xuanluo113 commented:
> @mjwitte And for the supplementary heater, I have to calculate the `FullOutPut` load because the `SupHeaterLoad` is based on the residual of `QZnReq - FullOutput`, and when that happens, it seems the `PartLoadFrac = 1.0; SpeedRatio = 1.0` and reevaluated `OnOffAirFlowRatio` should be local for the supplementary heat calc (as hard reset to simulate full load)?

Or should I use supplementary heater to make up for whatever load EMS override cannot met? I went back and forth for this piece.

xuanluo113 committed:
add EMS docs and modify Python API test file to be consistent

xuanluo113 commented:
@mjwitte I added the docs and modified the python API code. Running annually, the heating coil results make more sense.

<img width="616" alt="Screen Shot 2021-09-17 at 11 18 22 AM" src="https://user-images.githubusercontent.com/23508095/133835615-5e7ccaa2-329b-452d-9a12-4db526bc8b13.png">

The issues left are 
- sup heater
- error or warning message when EMS speed level > max speed level.

I'm actually leaning toward using sup heater to compliment whatever heating load EMS inputs cannot be met now, so I'll modify and test the logic now.

rraustad commented:
The fan mode does not matter. There is a load to meet. The supplemental heating coil should meet any load not met by the main heating coil up to the maximum supply air temperature. If the user does not want to meet the remianing load, the user can leave out the supplemental heating coil. I also see fatal errors if the speed choice does not match the load, not sure that is a good idea, EMS rules the roost.

xuanluo113 commented:
> The fan mode does not matter. There is a load to meet. The supplemental heating coil should meet any load not met by the main heating coil up to the maximum supply air temperature. If the user does not want to meet the remianing load, the user can leave out the supplemental heating coil. I also see fatal errors if the speed choice does not match the load, not sure that is a good idea, EMS rules the roost.

Could you elaborate on "fatal errors if the speed choice does not match the load"?

rraustad commented:
Sorry, I didn't word that correctly. I was referring to these fatal errors. The model should operate at full capacity if the EMS speed exceeds the number of coil speeds. SpeedNum=max and SpeedRatio = 1. If EMS speed = 5.3 I think you could just pass SpeedNum=5 to a 4 speed coil and it will operate at the max speed but you would want to opeate at SpeedRatio=1 not 0.3. And report a recurring warning so the user knows how often this happens.

        if (MSHeatPump(MSHeatPumpNum).HeatCoolMode == ModeOfOperation::HeatingMode) {
            if (SpeedNum > MSHeatPump(MSHeatPumpNum).NumOfSpeedHeating) {
                ShowFatalError(state,
                               "Wrong coil speed EMS override value, for unit=" + MSHeatPump(MSHeatPumpNum).DXCoolCoilName +
                                   ". Exceeding maximum coil speed level.");
            }
        } else if (MSHeatPump(MSHeatPumpNum).HeatCoolMode == ModeOfOperation::CoolingMode) {
            if (SpeedNum > MSHeatPump(MSHeatPumpNum).NumOfSpeedCooling) {
                ShowFatalError(state,

xuanluo113 committed:
add recurring warnings, reset maximum coil speed, and use tempout for sup heater

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

xuanluo113 commented:
@rraustad Thanks for the clarification. 

@rraustad @mjwitte In the latest commit,

- I added the recurring warning message to replace the fatal error. When a user inputs a value larger than maximum speed now, maximum speed is used and cycling/speed ratio is set to 1. Two unit tests are added to test this and the warming message in an annual run looks like
<img width="1059" alt="Screen Shot 2021-09-17 at 2 05 24 PM" src="https://user-images.githubusercontent.com/23508095/133853951-75ca6b4b-d460-4997-8f7f-f9f100c42c0b.png">

- Used the supplementary heater to fill as much unmet load of EMS control output as it can.

rraustad commented:
That recurring warning looks good. 
The recurring message says Max=3. Min=3.
Are you testing with a 2-speed coil?
If testing with a 2-speed coil then this is correct.
Just to be clear, I would expect Max=3 to mean you are testing with a 2-spd coil.

Question, why is the EMS program requesting SpeedNum=3? I guess because the zone temp far exceeds the set point.

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -8182,13 +8197,13 @@ namespace UnitarySystems {
                 this->m_HeatingSpeedRatio = 0.0;
                 this->m_HeatingCycRatio = this->m_EMSOverrideCoilSpeedNumValue - floor(this->m_EMSOverrideCoilSpeedNumValue);
-                if (this->m_HeatingCycRatio == 0) {
+                if (useMaxedSpeed or this->m_HeatingCycRatio == 0) {
```
Whoops! or should be ||

xuanluo113 replied:
😢  I've been switching between C++ and Python today.

rraustad commented on tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc:
```diff
@@ -1409,7 +1409,7 @@ TEST_F(EnergyPlusFixture, HVACMultiSpeedHeatPump_ReportVariableInitTest)
     EXPECT_NEAR(state->dataHVACMultiSpdHP->MSHeatPumpReport(2).CycRatio, 0.1530992, 0.0001);
     EXPECT_NEAR(state->dataHVACMultiSpdHP->MSHeatPump(2).CompPartLoadRatio, 0.1530992, 0.0001);
-    EXPECT_NEAR(state->dataLoopNodes->Node(22).Temp, 26.546664, 0.0001);
+    EXPECT_NEAR(state->dataLoopNodes->Node(22).Temp, 26.546664, 0.001);
```
No change required here but using this precision gets in trouble as things change over time. Near 0.1 or 0.01 is fine.

xuanluo113 commented:
> That recurring warning looks good.
> The recurring message says Max=3. Min=3.
> Are you testing with a 2-speed coil?
> If testing with a 2-speed coil then this is correct.
> Just to be clear, I would expect Max=3 to mean you are testing with a 2-spd coil.
> 
> Question, why is the EMS program requesting SpeedNum=3? I guess because the zone temp far exceeds the set point.

Yes. Max speed level is 2. And EMS is requesting a speed value of 2.5. (by hard coding in order to test the warning. This does not happen in the actual test file.)
I'm actually quite unsure about this. Shall I report the warning based on the evaluated speed level (e.g. `ceil(EMSSpeedValue) = 3) or just the user input speed value (so, 2.5, here).

xuanluo113 commented:
> > That recurring warning looks good.
> > The recurring message says Max=3. Min=3.
> > Are you testing with a 2-speed coil?
> > If testing with a 2-speed coil then this is correct.
> > Just to be clear, I would expect Max=3 to mean you are testing with a 2-spd coil.
> > Question, why is the EMS program requesting SpeedNum=3? I guess because the zone temp far exceeds the set point.
> 
> Yes. Max speed level is 2. And EMS is requesting a speed value of 2.5. (by hard coding in order to test the warning. This does not happen in the actual test file.)
> I'm actually quite unsure about this. Shall I report the warning based on the evaluated speed level (e.g. `ceil(EMSSpeedValue) = 3) or just the user input speed value (so, 2.5, here).

@rraustad By "report", I mean, the values of `ReportMaxOf` and `ReportMinOf` passed to `ShowRecurringWarningErrorAtEnd`. Or it's just not necessary to report the min and max values here to avoid confusion.

```
            ShowRecurringWarningErrorAtEnd(state,
                                           "Wrong coil speed EMS override value, for unit=\"" + useMaxedSpeedCoilName +
                                               "\". Exceeding maximum coil speed level. Speed level is set to the maximum coil speed level allowed.",
                                           MSHeatPump(MSHeatPumpNum).CoilSpeedErrIndex,
                                           ceil(SpeedVal),
                                           ceil(SpeedVal),
                                           _,
                                           "",
                                           "");
```

rraustad commented:
> Yes. Max speed level is 2. And EMS is requesting a speed value of 2.5. (by hard coding in order to test the warning. This does not happen in the actual test file.)
> I'm actually quite unsure about this. Shall I report the warning based on the evaluated speed level (e.g. `ceil(EMSSpeedValue) = 3) or just the user input speed value (so, 2.5, here).

If EMS is requesting 2.5 then the Max/Min statistics are misleading. Report the EMS speed request instead [user input speed value (so, 2.5, here)].

xuanluo113 committed:
or to actual or; report exact  min and max

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into coil-speed-actuator

rraustad commented:
This is getting close, maybe even good now.

xuanluo113 commented:
The unit test failed at the two commits ahead seems irrelevant to this branch.

Myoldmopar commented:
The one "failure" on Windows is the GPU issue that occurs on the new Windows Server CI instance, nothing to worry about.  @rraustad if you do feel this is good to go in, and CI comes back happy overnight, I can do some brief testing and get this in.

rraustad commented:
I will rely on @xuanluo0113 response to any of our comments. If this goes in to the release, and this in not correct, then we should have held it. I don't feel that at this point. @xuanluo113 ? any comments ?

xuanluo113 commented:
@rraustad Sorry I didn't make it clear before. In my latest commit, I've addressed the comments in our last discussion by reporting the exact min and max of speed value inputs (e.g. 2.5) instead of the rounding speed level. I think so far all discussions were addressed. I don't have anything else left behind that I'm aware of so far.

rraustad commented:
@xuanluo113 you were very clear. The thing is this branch will be merged with develop and we have to be sure it is correct. My review says this is good, CI may have issues, I was asking if you had any reservations of submitting this to the master respository? are you good?

xuanluo113 commented:
> @xuanluo113 you were very clear. The thing is this branch will be merged with develop and we have to be sure it is correct. My review says this is good, CI may have issues, I was asking if you had any reservations of submitting this to the master respository? are you good?

I don't have anything else left behind that I'm aware of so far. If CI turns green then I think it good to go.

rraustad commented:
I have seen this before so probably not related to this branch. The first time I run the unit tests locally I get a failure. Running again passes. And again, passes. And again, passes. So unit tests do run successfully.

Myoldmopar commented:
This still looks perfect locally, and CI is super happy.  Let's get this one out of the way.  Thanks @xuanluo113 and the reviewers!
